File: examples/ch10/ts/conceptSingleton.ts
// Singleton — canonical conceptual example. Run: npx tsx conceptSingleton.ts
//
// The classical Singleton ensures one instance with global access. The
// idiomatic form needs none of it: an ES module is itself a singleton —
// evaluated once and cached — so a module-level instance gives the same
// guarantee and stays injectable for tests.

// --- Canonical: the Singleton pattern (rarely needed) -----------------------

export class Singleton {
  private static instance: Singleton | undefined;

  // A private constructor blocks `new Singleton()` from outside.
  private constructor() {}

  static getInstance(): Singleton {
    if (Singleton.instance === undefined) {
      Singleton.instance = new Singleton();
    }
    return Singleton.instance;
  }
}

// --- Idiomatic: a module-level instance (this object, imported, IS the singleton)

export class Config {
  theme = "light";
}

export const config = new Config(); // created once at module load; importable everywhere

// Demo
console.log(Singleton.getInstance() === Singleton.getInstance()); // true — but it took machinery
console.log(config === config); // true — and this needed none

File: examples/ch10/py/concept_singleton.py
"""Singleton — canonical conceptual example (runnable).

The classical Singleton ensures one instance with global access. The Pythonic
form needs none of it: a module is already a singleton, so a module-level
instance gives the same guarantee and stays injectable for tests.
"""

from __future__ import annotations


# --- Canonical: the Singleton pattern ---------------------------------------

class Singleton:
    _instance: Singleton | None = None

    def __new__(cls) -> Singleton:
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance


# --- Pythonic: a module-level instance (this object, imported, IS the singleton)

class Config:
    def __init__(self) -> None:
        self.theme = "light"


config = Config()   # created once at import; `from concept_singleton import config`


if __name__ == "__main__":
    print(Singleton() is Singleton())   # True — but it took a __new__ to get here
    print(config is config)             # True — and this needed no machinery
